### ISSUE TITLE
logoSvg attribute of endpoint does not function

### ISSUE THREAD

**Issue by FerreTux** [badges/shields#6335]
Are you experiencing an issue with...

- [x ] [shields.io](https://shields.io/#/)
- [ ] My own instance
- [ ] [badge-maker NPM package](https://www.npmjs.com/package/badge-maker)

:beetle: **Description**

When loading badges with endpoint the site says the attribute logoSvg is supported in the json payload.  From other areas of the site and digging in the code a bit it implies this must be a sorted base64 SVG(preferred) or a png <8192 bytes.   I tried some known good svgs from other areas of the site and loaded them into the payload and still no dice getting endpoint to display the custom logo.    I have tried the following values for logoSvg as well as url safe variations and a number of other svg's and variations on how to pass the data in the payload.    I am at a loss at this point

## My test custom image 1
data:image/svg+xml;base64,PHN2ZyB4bWxucz0iaHR0cDovL3d3dy53My5vcmcvMjAwMC9zdmciIGRhdGEtbmFtZT0iTGF5ZXIgMSIgdmlld0JveD0iMCAwIDczLjA5IDczLjA5Ij48ZGVmcz48bGluZWFyR3JhZGllbnQgaWQ9ImEiIHgxPSI0LjkyMyIgeDI9IjY4LjExMyIgeTE9IjM2LjUxOCIgeTI9IjM2LjUxOCIgZ3JhZGllbnRVbml0cz0idXNlclNwYWNlT25Vc2UiPjxzdG9wIG9mZnNldD0iMCIgc3RvcC1jb2xvcj0iI2ZmZiIvPjxzdG9wIG9mZnNldD0iLjQ4MyIgc3RvcC1jb2xvcj0iIzY2NiIvPjxzdG9wIG9mZnNldD0iLjU3MyIgc3RvcC1jb2xvcj0iIzYxNjE2MSIvPjxzdG9wIG9mZnNldD0iLjY4MiIgc3RvcC1jb2xvcj0iIzUyNTI1MiIvPjxzdG9wIG9mZnNldD0iLjgwMiIgc3RvcC1jb2xvcj0iIzM5MzkzOSIvPjxzdG9wIG9mZnNldD0iLjkyOCIgc3RvcC1jb2xvcj0iIzE3MTcxNyIvPjxzdG9wIG9mZnNldD0iMSIvPjwvbGluZWFyR3JhZGllbnQ-PC9kZWZzPjxwYXRoIGZpbGw9InVybCgjYSkiIHN0cm9rZT0iIzAwMCIgc3Ryb2tlLW1pdGVybGltaXQ9IjEwIiBzdHJva2Utd2lkdGg9IjciIGQ9Ik00LjkyMyAzNi4yMTJsMzEuMjIgMzEuOTAxIDMxLjk3LTMxLjk3LTMxLjkwMS0zMS4yMkw0LjkyMyAzNi4yMTJ6Ii8-PC9zdmc-

## My Custom image 2
data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAAEoAAABKCAMAAAArDjJDAAABpFBMVEX///8AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAGBgYHBwcICAgJCQkKCgoLCwsMDAwNDQ0ODg4RERESEhITExMUExMUFBQVFRUWFBUWFhYXFhYXFxcYGBgZGBgZGRkaGRkaGhobGxscGxscHBwdHR0eHh4fHx8gHx8gICAhICAhISEiICAiIiIjIyMkIiMkJCQlJCQlJSUmJCUmJiYnJic6Nzg9Ojs/PT5CQEFFQ0NHRUZKSElMS0tOTU1RUFBTUlJUVFRVVFRXVlZYV1dYWFhZWFhZWVlaWlpbWlpcW1tcW1xcXFxdXF1eXV1eXl5fX19gYGBhYWFiYWFjYmNkY2NkZGRlZWVmZmZoaGhpaWlqampsbGxtbW1wcHBxcXFzc3N1dXV3d3d4eHh5eXl7e3t8fHx9fX1/f3+BgYGCgoKEhISFhYWIiIiJiYmKioqMjIyNjY2Pj4+QkJCUlJSVlZWZmZmdnZ2fn5+ioqKjo6OmpqaoqKiqqqqurq6ysrK2tra3t7e6urq9vb3BwcHHx8fMzMzR0dHW1tbb29vg4OBa4AxFAAAACnRSTlMAEDBAYJCg0ODw5HV5kgAAAf5JREFUeNqt1gOXazEUBeBaz7Zt23ZtPtt2O9aPHk/2uU5yupcuv6Uc+GTi9/t4gRSNMi1IiUQs0C0pkYizLEhcCxLXgsS1IPEtSHwLEt+CNG8lLKZ0NcuyIM2/Wa1QiyE9aFYrlcwKYQW1pdzDKaqcFlYipCctKD2ZoUopYmlJtedzVDFJLA2p8RpU4R6xlKXWO0rl7yhakBY+/miksreXw1KSnn41U5kbxFKQXnyzUqn7B2HJSote/7Cl7u6HJSm9/eVA3doLS0p6/8eRur4HloT0+Z8LdW23sMJe0uKvbVfqyi5hRTyk7x0P6vJOWK7Sj15P6tIOWM7Skp/9EtTF7bAcpd8DUtSFrbAcpL+DktT5LVaLSkv/D0lT5/ZZLSK1hxWosydhWaTOiBJ15jgsk9QzqkidPmywhLSsb0yZOnWAWhFx80WDWkVLm1iflKm1BolaHxSp9SaJWu+UqA0WiVpvFKhNNhK1XmkUDpWopVjOkGDpNhlIsDRbX4jfkCHxxwQk/vCC1P2R2v1Bz18/mEsRY1VLUYm1QGYO6S6j216bKEWJWptfGqgj0hKsmLCeEeqovIQE4nM/bXwkqGM6ErU2tGYpTYla6+rTlLZErTWVSeqEvkSt1TmeRK2VTIlaXAkWX4LFl2CxJFiQmAlCYifEl2DxJVh8CZaUNAGUeEDUbbHUbwAAAABJRU5ErkJggg==

## Known good
data:image/svg+xml;base64,PHN2ZyB4bWxucz0iaHR0cDovL3d3dy53My5vcmcvMjAwMC9zdmciIHZlcnNpb249IjEiIHdpZHRoPSI2MDAiIGhlaWdodD0iNjAwIj48cGF0aCBkPSJNMTI5IDExMWMtNTUgNC05MyA2Ni05MyA3OEwwIDM5OGMtMiA3MCAzNiA5MiA2OSA5MWgxYzc5IDAgODctNTcgMTMwLTEyOGgyMDFjNDMgNzEgNTAgMTI4IDEyOSAxMjhoMWMzMyAxIDcxLTIxIDY5LTkxbC0zNi0yMDljMC0xMi00MC03OC05OC03OGgtMTBjLTYzIDAtOTIgMzUtOTIgNDJIMjM2YzAtNy0yOS00Mi05Mi00MmgtMTV6IiBmaWxsPSIjZmZmIi8+PC9zdmc+

# Worth noting that some of these work from url logo field
![Known Good](https://img.shields.io/badge/play-station-blue.svg?logo=data:image/svg%2bxml;base64,PHN2ZyB4bWxucz0iaHR0cDovL3d3dy53My5vcmcvMjAwMC9zdmciIHZlcnNpb249IjEiIHdpZHRoPSI2MDAiIGhlaWdodD0iNjAwIj48cGF0aCBkPSJNMTI5IDExMWMtNTUgNC05MyA2Ni05MyA3OEwwIDM5OGMtMiA3MCAzNiA5MiA2OSA5MWgxYzc5IDAgODctNTcgMTMwLTEyOGgyMDFjNDMgNzEgNTAgMTI4IDEyOSAxMjhoMWMzMyAxIDcxLTIxIDY5LTkxbC0zNi0yMDljMC0xMi00MC03OC05OC03OGgtMTBjLTYzIDAtOTIgMzUtOTIgNDJIMjM2YzAtNy0yOS00Mi05Mi00MmgtMTV6IiBmaWxsPSIjZmZmIi8+PC9zdmc+)

![Test Case 2](https://img.shields.io/badge/test-case2-blue.svg?logo=data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAAEoAAABKCAMAAAArDjJDAAABpFBMVEX///8AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAGBgYHBwcICAgJCQkKCgoLCwsMDAwNDQ0ODg4RERESEhITExMUExMUFBQVFRUWFBUWFhYXFhYXFxcYGBgZGBgZGRkaGRkaGhobGxscGxscHBwdHR0eHh4fHx8gHx8gICAhICAhISEiICAiIiIjIyMkIiMkJCQlJCQlJSUmJCUmJiYnJic6Nzg9Ojs/PT5CQEFFQ0NHRUZKSElMS0tOTU1RUFBTUlJUVFRVVFRXVlZYV1dYWFhZWFhZWVlaWlpbWlpcW1tcW1xcXFxdXF1eXV1eXl5fX19gYGBhYWFiYWFjYmNkY2NkZGRlZWVmZmZoaGhpaWlqampsbGxtbW1wcHBxcXFzc3N1dXV3d3d4eHh5eXl7e3t8fHx9fX1/f3+BgYGCgoKEhISFhYWIiIiJiYmKioqMjIyNjY2Pj4+QkJCUlJSVlZWZmZmdnZ2fn5+ioqKjo6OmpqaoqKiqqqqurq6ysrK2tra3t7e6urq9vb3BwcHHx8fMzMzR0dHW1tbb29vg4OBa4AxFAAAACnRSTlMAEDBAYJCg0ODw5HV5kgAAAf5JREFUeNqt1gOXazEUBeBaz7Zt23ZtPtt2O9aPHk/2uU5yupcuv6Uc+GTi9/t4gRSNMi1IiUQs0C0pkYizLEhcCxLXgsS1IPEtSHwLEt+CNG8lLKZ0NcuyIM2/Wa1QiyE9aFYrlcwKYQW1pdzDKaqcFlYipCctKD2ZoUopYmlJtedzVDFJLA2p8RpU4R6xlKXWO0rl7yhakBY+/miksreXw1KSnn41U5kbxFKQXnyzUqn7B2HJSote/7Cl7u6HJSm9/eVA3doLS0p6/8eRur4HloT0+Z8LdW23sMJe0uKvbVfqyi5hRTyk7x0P6vJOWK7Sj15P6tIOWM7Skp/9EtTF7bAcpd8DUtSFrbAcpL+DktT5LVaLSkv/D0lT5/ZZLSK1hxWosydhWaTOiBJ15jgsk9QzqkidPmywhLSsb0yZOnWAWhFx80WDWkVLm1iflKm1BolaHxSp9SaJWu+UqA0WiVpvFKhNNhK1XmkUDpWopVjOkGDpNhlIsDRbX4jfkCHxxwQk/vCC1P2R2v1Bz18/mEsRY1VLUYm1QGYO6S6j216bKEWJWptfGqgj0hKsmLCeEeqovIQE4nM/bXwkqGM6ErU2tGYpTYla6+rTlLZErTWVSeqEvkSt1TmeRK2VTIlaXAkWX4LFl2CxJFiQmAlCYifEl2DxJVh8CZaUNAGUeEDUbbHUbwAAAABJRU5ErkJggg==)

:link: **Link to the badge**
## Test Case 1
![ferretuxLogo](https://img.shields.io/endpoint?url=https://gist.githubusercontent.com/FerreTux/761627e5ad10843ebc983328034a8e3f/raw/ferretuxLogo)

## Test Case 2

![ferretuxLogo](https://img.shields.io/endpoint?url=https://gist.githubusercontent.com/FerreTux/f3a43cda271a91900d1296a571bfb522/raw/bce8bb7051a67ecb780f5660647f3f86ab67c3a9/Test%2520Case%25202)

## Known Good
![ferretuxLogo](https://img.shields.io/endpoint?url=https://gist.githubusercontent.com/FerreTux/f3a43cda271a91900d1296a571bfb522/raw/bce8bb7051a67ecb780f5660647f3f86ab67c3a9/Known%2520Good)

:bulb: **Possible Solution**

My guess is this is related to the packages that are outdated.

But I haven't  ruled out 
- The images are not quite in the right format and the system is rejecting them because its super picky
- This attribute isn't implemented

Ultimately I think shields needs a much better explanation of how to use endpoint fields with much more verbose errors dumped to the end user when things go wrong.  The caching can be a nightmare to work with when testing.  Would be nice to have a debug flag or something to force caching to 0 or something.   It would be especially nice if the system is picky about formats to do the svg/png conversion directly on the site for folks.

**Comment by calebcartwright** [comment]
It sounds like you're pretty frustrated, sorry to hear that!

Seems you've confused the `logo` query parameter on our badge urls  (which is a base64 encoded value) with the `logoSvg` field in the Endpoint badge response schema. You're trying to use the latter but with the format of the former, and they're not compatible.

For the `logoSvg` key in your Endpoint response, you need to provide an SVG string (e.g. `"logoSvg": "<svg>....</svg>"`) for the value. From the Endpoint schema documentation (https://shields.io/endpoint):
> logoSvg     Default: none.      An SVG string containing a custom logo.

Whereas for the `logo` url query parameter (e.g. `?logo=data:image/svg+xml;base64,PHN2Z...`) you need a base64 value as noted in both our in-repo documentation on custom logos (https://github.com/badges/shields/blob/master/doc/logos.md#custom-logos) as well as the the Shields.io documentation for the param https://shields.io/#styles
> Insert custom logo image (≥ 14px high). There is a limit on the total size of request headers we can accept (8192 bytes). From a practical perspective, this means the base64-encoded image text is limited to somewhere slightly under 8192 bytes depending on the rest of the request header.

If you have any concrete suggestions on improvements to the documentation though, please do let us know, or even submit a PR! 

Now to respond directly to a few of your points:

> My test custom image 1

As discussed above, the core problem here is the invalid value in your Endpoint response, but more generally are you sure this is a valid base64 encoded string? Both the raw value you posted in the issue description and the (identical) value you've set in the gist file have `-` characters and perhaps I am remembering incorrectly, but I don't believe `-` are valid/allowed base64 chars.

I think this is reflected in the fact that encoded value as provided doesn't work anywhere:

https://img.shields.io/badge/foo-bar-blue?logo=data:image/svg+xml;base64,PHN2ZyB4bWxucz0iaHR0cDovL3d3dy53My5vcmcvMjAwMC9zdmciIGRhdGEtbmFtZT0iTGF5ZXIgMSIgdmlld0JveD0iMCAwIDczLjA5IDczLjA5Ij48ZGVmcz48bGluZWFyR3JhZGllbnQgaWQ9ImEiIHgxPSI0LjkyMyIgeDI9IjY4LjExMyIgeTE9IjM2LjUxOCIgeTI9IjM2LjUxOCIgZ3JhZGllbnRVbml0cz0idXNlclNwYWNlT25Vc2UiPjxzdG9wIG9mZnNldD0iMCIgc3RvcC1jb2xvcj0iI2ZmZiIvPjxzdG9wIG9mZnNldD0iLjQ4MyIgc3RvcC1jb2xvcj0iIzY2NiIvPjxzdG9wIG9mZnNldD0iLjU3MyIgc3RvcC1jb2xvcj0iIzYxNjE2MSIvPjxzdG9wIG9mZnNldD0iLjY4MiIgc3RvcC1jb2xvcj0iIzUyNTI1MiIvPjxzdG9wIG9mZnNldD0iLjgwMiIgc3RvcC1jb2xvcj0iIzM5MzkzOSIvPjxzdG9wIG9mZnNldD0iLjkyOCIgc3RvcC1jb2xvcj0iIzE3MTcxNyIvPjxzdG9wIG9mZnNldD0iMSIvPjwvbGluZWFyR3JhZGllbnQ-PC9kZWZzPjxwYXRoIGZpbGw9InVybCgjYSkiIHN0cm9rZT0iIzAwMCIgc3Ryb2tlLW1pdGVybGltaXQ9IjEwIiBzdHJva2Utd2lkdGg9IjciIGQ9Ik00LjkyMyAzNi4yMTJsMzEuMjIgMzEuOTAxIDMxLjk3LTMxLjk3LTMxLjkwMS0zMS4yMkw0LjkyMyAzNi4yMTJ6Ii8-PC9zdmc-
![](https://img.shields.io/badge/foo-bar-blue?logo=data:image/svg+xml;base64,PHN2ZyB4bWxucz0iaHR0cDovL3d3dy53My5vcmcvMjAwMC9zdmciIGRhdGEtbmFtZT0iTGF5ZXIgMSIgdmlld0JveD0iMCAwIDczLjA5IDczLjA5Ij48ZGVmcz48bGluZWFyR3JhZGllbnQgaWQ9ImEiIHgxPSI0LjkyMyIgeDI9IjY4LjExMyIgeTE9IjM2LjUxOCIgeTI9IjM2LjUxOCIgZ3JhZGllbnRVbml0cz0idXNlclNwYWNlT25Vc2UiPjxzdG9wIG9mZnNldD0iMCIgc3RvcC1jb2xvcj0iI2ZmZiIvPjxzdG9wIG9mZnNldD0iLjQ4MyIgc3RvcC1jb2xvcj0iIzY2NiIvPjxzdG9wIG9mZnNldD0iLjU3MyIgc3RvcC1jb2xvcj0iIzYxNjE2MSIvPjxzdG9wIG9mZnNldD0iLjY4MiIgc3RvcC1jb2xvcj0iIzUyNTI1MiIvPjxzdG9wIG9mZnNldD0iLjgwMiIgc3RvcC1jb2xvcj0iIzM5MzkzOSIvPjxzdG9wIG9mZnNldD0iLjkyOCIgc3RvcC1jb2xvcj0iIzE3MTcxNyIvPjxzdG9wIG9mZnNldD0iMSIvPjwvbGluZWFyR3JhZGllbnQ-PC9kZWZzPjxwYXRoIGZpbGw9InVybCgjYSkiIHN0cm9rZT0iIzAwMCIgc3Ryb2tlLW1pdGVybGltaXQ9IjEwIiBzdHJva2Utd2lkdGg9IjciIGQ9Ik00LjkyMyAzNi4yMTJsMzEuMjIgMzEuOTAxIDMxLjk3LTMxLjk3LTMxLjkwMS0zMS4yMkw0LjkyMyAzNi4yMTJ6Ii8-PC9zdmc-)

> My Custom image 2

As shared above and as the `logoSvg` field name and documentation suggest, this needs to be a valid svg string, not a base64 encoded png. If you don't want to convert this to an SVG, then you'd need to tack on the `&logo=...` with the base64 value to your badge urls:

![](https://img.shields.io/endpoint?url=https://gist.githubusercontent.com/FerreTux/f3a43cda271a91900d1296a571bfb522/raw/bce8bb7051a67ecb780f5660647f3f86ab67c3a9/Test%2520Case%25202&logo=data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAAEoAAABKCAMAAAArDjJDAAABpFBMVEX///8AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAGBgYHBwcICAgJCQkKCgoLCwsMDAwNDQ0ODg4RERESEhITExMUExMUFBQVFRUWFBUWFhYXFhYXFxcYGBgZGBgZGRkaGRkaGhobGxscGxscHBwdHR0eHh4fHx8gHx8gICAhICAhISEiICAiIiIjIyMkIiMkJCQlJCQlJSUmJCUmJiYnJic6Nzg9Ojs/PT5CQEFFQ0NHRUZKSElMS0tOTU1RUFBTUlJUVFRVVFRXVlZYV1dYWFhZWFhZWVlaWlpbWlpcW1tcW1xcXFxdXF1eXV1eXl5fX19gYGBhYWFiYWFjYmNkY2NkZGRlZWVmZmZoaGhpaWlqampsbGxtbW1wcHBxcXFzc3N1dXV3d3d4eHh5eXl7e3t8fHx9fX1/f3+BgYGCgoKEhISFhYWIiIiJiYmKioqMjIyNjY2Pj4+QkJCUlJSVlZWZmZmdnZ2fn5+ioqKjo6OmpqaoqKiqqqqurq6ysrK2tra3t7e6urq9vb3BwcHHx8fMzMzR0dHW1tbb29vg4OBa4AxFAAAACnRSTlMAEDBAYJCg0ODw5HV5kgAAAf5JREFUeNqt1gOXazEUBeBaz7Zt23ZtPtt2O9aPHk/2uU5yupcuv6Uc+GTi9/t4gRSNMi1IiUQs0C0pkYizLEhcCxLXgsS1IPEtSHwLEt+CNG8lLKZ0NcuyIM2/Wa1QiyE9aFYrlcwKYQW1pdzDKaqcFlYipCctKD2ZoUopYmlJtedzVDFJLA2p8RpU4R6xlKXWO0rl7yhakBY+/miksreXw1KSnn41U5kbxFKQXnyzUqn7B2HJSote/7Cl7u6HJSm9/eVA3doLS0p6/8eRur4HloT0+Z8LdW23sMJe0uKvbVfqyi5hRTyk7x0P6vJOWK7Sj15P6tIOWM7Skp/9EtTF7bAcpd8DUtSFrbAcpL+DktT5LVaLSkv/D0lT5/ZZLSK1hxWosydhWaTOiBJ15jgsk9QzqkidPmywhLSsb0yZOnWAWhFx80WDWkVLm1iflKm1BolaHxSp9SaJWu+UqA0WiVpvFKhNNhK1XmkUDpWopVjOkGDpNhlIsDRbX4jfkCHxxwQk/vCC1P2R2v1Bz18/mEsRY1VLUYm1QGYO6S6j216bKEWJWptfGqgj0hKsmLCeEeqovIQE4nM/bXwkqGM6ErU2tGYpTYla6+rTlLZErTWVSeqEvkSt1TmeRK2VTIlaXAkWX4LFl2CxJFiQmAlCYifEl2DxJVh8CZaUNAGUeEDUbbHUbwAAAABJRU5ErkJggg==
)

https://img.shields.io/endpoint?url=https://gist.githubusercontent.com/FerreTux/f3a43cda271a91900d1296a571bfb522/raw/bce8bb7051a67ecb780f5660647f3f86ab67c3a9/Test%2520Case%25202?logo=data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAAEoAAABKCAMAAAArDjJDAAABpFBMVEX///8AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAGBgYHBwcICAgJCQkKCgoLCwsMDAwNDQ0ODg4RERESEhITExMUExMUFBQVFRUWFBUWFhYXFhYXFxcYGBgZGBgZGRkaGRkaGhobGxscGxscHBwdHR0eHh4fHx8gHx8gICAhICAhISEiICAiIiIjIyMkIiMkJCQlJCQlJSUmJCUmJiYnJic6Nzg9Ojs/PT5CQEFFQ0NHRUZKSElMS0tOTU1RUFBTUlJUVFRVVFRXVlZYV1dYWFhZWFhZWVlaWlpbWlpcW1tcW1xcXFxdXF1eXV1eXl5fX19gYGBhYWFiYWFjYmNkY2NkZGRlZWVmZmZoaGhpaWlqampsbGxtbW1wcHBxcXFzc3N1dXV3d3d4eHh5eXl7e3t8fHx9fX1/f3+BgYGCgoKEhISFhYWIiIiJiYmKioqMjIyNjY2Pj4+QkJCUlJSVlZWZmZmdnZ2fn5+ioqKjo6OmpqaoqKiqqqqurq6ysrK2tra3t7e6urq9vb3BwcHHx8fMzMzR0dHW1tbb29vg4OBa4AxFAAAACnRSTlMAEDBAYJCg0ODw5HV5kgAAAf5JREFUeNqt1gOXazEUBeBaz7Zt23ZtPtt2O9aPHk/2uU5yupcuv6Uc+GTi9/t4gRSNMi1IiUQs0C0pkYizLEhcCxLXgsS1IPEtSHwLEt+CNG8lLKZ0NcuyIM2/Wa1QiyE9aFYrlcwKYQW1pdzDKaqcFlYipCctKD2ZoUopYmlJtedzVDFJLA2p8RpU4R6xlKXWO0rl7yhakBY+/miksreXw1KSnn41U5kbxFKQXnyzUqn7B2HJSote/7Cl7u6HJSm9/eVA3doLS0p6/8eRur4HloT0+Z8LdW23sMJe0uKvbVfqyi5hRTyk7x0P6vJOWK7Sj15P6tIOWM7Skp/9EtTF7bAcpd8DUtSFrbAcpL+DktT5LVaLSkv/D0lT5/ZZLSK1hxWosydhWaTOiBJ15jgsk9QzqkidPmywhLSsb0yZOnWAWhFx80WDWkVLm1iflKm1BolaHxSp9SaJWu+UqA0WiVpvFKhNNhK1XmkUDpWopVjOkGDpNhlIsDRbX4jfkCHxxwQk/vCC1P2R2v1Bz18/mEsRY1VLUYm1QGYO6S6j216bKEWJWptfGqgj0hKsmLCeEeqovIQE4nM/bXwkqGM6ErU2tGYpTYla6+rTlLZErTWVSeqEvkSt1TmeRK2VTIlaXAkWX4LFl2CxJFiQmAlCYifEl2DxJVh8CZaUNAGUeEDUbbHUbwAAAABJRU5ErkJggg==

> Ultimately I think shields needs a much better explanation of how to use endpoint fields 

I think this has been covered above already, but will reiterate that any concrete suggestions/improvements to the docs that might help future users from making the same mistake will definitely be appreciated!

> with much more verbose errors dumped to the end user when things go wrong.

I'm not really sure what you have in mind here, especially since we don't really have a means to "dump" anything to users outside the textual values that go into the svg we return and associated response headers.

> The caching can be a nightmare to work with when testing. Would be nice to have a debug flag or something to force caching to 0 or something

I do not follow the relevance of caching to the issue/confusion here, nor do understand how allowing users to blow away all of our caching would have been of assistance. Chris provided an excellent summary of caching in a recent issue that I'll simply [point to](https://github.com/badges/shields/issues/6310#issuecomment-807387522) instead of repeating here, but a core take away is that caching is critical to the reliability and availability of the service. We do get our fair share of folks trying to DDOS us, so we have to keep the big picture in mind and weigh the tradeoffs.

> It would be especially nice if the system is picky about formats to do the svg/png conversion directly on the site for folks.

This has been brought up before, but our response remains the same: we're not going to do this. There's myriad free online and command line tools available to anyone that can be used for such conversions. We can understand how some users feel it might be a convenience for Shields.io to provide yet-another-such option for one-time conversions, and/or for our badge servers to support doing that on the fly as part of serving badges. However, the drawbacks from a maintenance, runtime overhead, etc. of us doing this significantly outweigh the minor user convenience, and it's only a minor inconvenience for users to do this themselves with the available tooling, especially since it's a one time action for them.

**Comment by FerreTux** [comment]
Thanks for the thorough response,  I swear I had tried that before but it was late at night when I did. so entirely possibly  screwed it up then and never went back to try again.  I did get my custom logo to finally work but I had to remove schema version from the payload.  Which is a bit ironic considering its required field,  but that just may have been from the small endpoint tester at bottom of page not liking it, Or a great concrete example of how the documentation could be better.  

As for concrete example of how documentation could be better. [Random notes](https://docs.google.com/document/d/1f7Zf5gWvKq-F3hmD6dxmS77adh7N_TBqfthQxgzo21k/edit?usp=sharing)

In general the documentation feels really "strewn across the floors a bit."  please don't take that the wrong way.   Most of my gripes relate to being a newcomer to using badges and shields,io in particular.  I was absolutely blown away that the hardest part of my project ended up being getting the badges working and not the code supporting it.   I literally learned github actions and workflows faster than making a badge which is unfortunate..... to say the least.   


As for how caching can be a pain.   If you are making tweaks and trying to see the results waiting 5 mins between json changes really is a long time.  I eventually learned the little tester at bottom of the page was less prone to caching delays between tests but that wasn't immediately clear,  probably because I was working on this late.

**Comment by calebcartwright** [comment]
> please don't take that the wrong way

To be frank, your choice of wording throughout this issue is pretty clearly conveying some bitterness and frustration. I think you'll find folks here, both maintainers and the community, are happy and willing to help out and there's a lot of ways of getting in contact with us. I would suggest for any future issues, particularly if you get stuck, that you consider reaching out earlier for tips/assistance/etc. vs. going it alone, getting frustrated, and only afterwards coming forward and conveying said frustration.

The Endpoint badge is for relatively advanced use cases, and not what most people use (most just go to Shields.io, search/browse for their service or badge, plug in their project details, and call it a day). As such, it would be great to get an understanding of why your very first badge experience required jumping straight to the Endpoint badge, and why the Endpoint schema documentation (or lackthereof) sent you down a path of incorrectly trying to use the base64 encoded strings from the `logo` queryParam as the values for the `logoSvg` key in your endpoint response. I'm unclear on how exactly that transpired, but would love to make any improvements we can on our end to help prevent that happening to anyone else.

> As for concrete example of how documentation could be better. Random notes

Thank you for taking the time to capture your thoughts and screenshots. 

While feedback on general UX is helpful and appreciated, I'm going to skip those stylistic/aesthetic pieces because I feel those are fairly subjective and orthogonal to the documentation/preventing others from mistaking the `logo` query param for the `logoSvg` endpoint field.

For some of the other items in your doc (my summary/wording in some cases to reflect text-in-screenshots)

> schemaVersion labeled as required, but not actually required

We deliberately call this out as required in the documentation because we very much want folks to add this. We retain the right to make changes to the endpoint badge at any point in time, and if a user _hasn't_ explicitly specified the schema version in their Endpoint then there is a chance they could see their badges break when we make changes. They'd then have to go back in and add `schemaVersion` retroactively and/or update their endpoint response structure to comply with the new version. By setting the `schemaVersion` in the Endpoint response, those badges are then protected from breaking schema changes.

However, it is indeed true that we do not fail/refuse to render Endpoint badges that elect to elide this value. This was also an intentional design decision.

> isError field bugs you/don't understand what the point is

The endpoint badges are primarily geared towards providing a plugin architecture of sorts that allows the Shields platform to provide native badges for various scenarios where we couldn't (or wouldn't) pull the implementation directly into our codebase along with the [other services](https://github.com/badges/shields/tree/master/services) we support natively. However, in many ways, Endpoints are an extension of the Shields platform, and support some cases which need to be able to utilize some core hooks of our internal badgel infrastructure. Being able to designate an error scenario is an important part of that for badge data fetchers & producers, and Endpoints inherently are just like these. The Endpoint schema is designed to support a variety of use cases, but that doesn't mean every use case needs to use every part of it, just as many other Endpoint badges don't need to utilize `logoSvg` in theirs.

> logoSvg suggestion:
```
Add 3 additional data points
1. Do not add schemaVersion when using logoSvg
2. logo should be less than < 8192 bytes and at least 14x14
3. Format should be "<svg>...</svg> with escaped quotes"
``` 

My responses:
1) It absolutely should not be the case that using both `schemaVersion` and `logoSvg` would be problematic, and would unequivocally be a bug if you are experiencing this. Can you please provide a reproduction case where an endpoint badge providing a response with a valid/correct value for `logoSvg` does not work when `schemaVersion` is set, as well as the corresponding one where that same response sans `schemaVersion` does work?
2) I don't believe this is strictly true, and feel you are again mixing the `logo` badge url query param and the `logoSvg` field in the endpoint schema. The constraints with the size actually comes from Node.js itself which sets a ceiling on the size of parts of the received request. That's applicable for `logo` query param because that's part of the _request_ that our badge servers receive, but that wouldn't necessarily be an issue with the Endpoint badge because it's in the _body_ of the _response_
3) I agree that perhaps an example could be helpful. However, important to note that the Shields.io service is explicitly and unapologetically targeted to, and built for, developers. As such we very much make some assumptions about our users having some core/foundational knowledge, including how to properly craft json. We do not view it as our responsibility to remind nor educate folks on some of these foundational things, including the need/how to escape things within strings in json.

> A Json Test Bed

This is an interesting idea, and I have some vague memories of discussions on a similar topic a while ago. I'm not necessarily opposed to this personally, but I also don't view it as a terribly high priority. The type of capabilities something like this would provide include things like:

* helping users craft valid json, escaping things on their behalf
* validating icon names from simple-icons
* performing svg validation

All of which are again niceties, but adds work on our plate that's not necessarily related to our core mission and all of which can be pretty easily handled by developers with existing tooling.

> Additional side note:  would be nice to support something like logoSvgUrl in the event you have it hosted somewhere that way you don’t have to update your payloads each time you want to ref your logo.

I spoke to this in my previous response, it's not something we're going to do. Endpoint badges exist so developers can do programmatic things outside of the running Shields server itself. Developers of Endpoint badges can easily program their endpoint to fetch a URL if that's something that's important for their use case. It doesn't really make sense for an Endpoint to toss a URL to us and ask us to do it instead.

> As for how caching can be a pain. If you are making tweaks and trying to see the results waiting 5 mins between json changes really is a long time. I eventually learned the little tester at bottom of the page was less prone to caching delays between tests but that wasn't immediately clear, probably because I was working on this late.

I'd suggest that much of this was probably driven by the fact that you'd made an incorrect diagnosis and were attempting solutions based on that diagnosis. It wouldn't have mattered whether the cache was 0 or 24 hours so long as you were trying things like an invalid value in `logoSvg` with a base64 png in your response.

You can also often utilize query parameters on the badge url in these types of scenarios too (you can still add `?label=foo` to the badge url which will override the value in the endpoint response, and you can iterate this as needed `?label=bar`, etc. This will let you skip the cache)

**Comment by FerreTux** [comment]
**Commentary and tone of discussion**
I apologize if anything came off that way., that was not intended **AT ALL**; pitfalls of textual communication =p.   I only meant to point out that something could be improved with the documentation.   I probably just got **WAY** to comfortable over the last 5 years working with a tight nit team.  Easy to forget that brutal honest isn't a great opener.    

In general, this explains why I thought your initial response was a bit defensive.   Again I apologize if you felt that way and you are completely and totally right. **I need to interact with other developers a lot more and not go it alone always.** Seriously heart felt I know.   That's really what I have been working towards this last few weeks and why I chose to jump back on GitHub and start working on stuff again.  

 **why endpoint?**
    A lot of the Actions I was finding on Github for badges were not very well dynamic.  at best they could set a single badge at a time.  I wanted something more dynamic that could both function statically and dynamically and still be easily accessible from within the project IDE on my local machine.   After not finding something to do exactly what I wanted and reading comments in a lot of forums from other developers wanting similar things,  I figured why not take on the simple challenge of writing a custom action in Github to do all that, leveraging your service.    Even if there is something else out there,  it was still something good to get me back into the habit of coding things again after spending a few years managing a tech team at Google.   I threw all my knowledge set at learning their systems etc. anything else is like reading ancient Babylonian its been so long since touching things non-Google.  We were probably more than a little spoiled.
     Anyways,  I set out to make something that just takes one large json file, splits it and posts it to independent gist files so all my badges get updated in a single pass.   I figured this would be helpful for later if I wanted to have one application run through tests etc and its output could then be easily sent as a single payload without to much headache.   So endpoint was the perfect thing to work without coding all the styling myself. 
    
**Development / Recommendations**     
I do very much like your service and completely understand its unlikely to be a priority.  Your not Google where UX is an obsession

I also understand that good UX shouldn't be sacrificed just because there is assumed skillsets knowledge.  After all developers like nice things too ;) else badges, linters, unit testers wouldn't exist.   

```
Add 3 additional data points
1. Do not add schemaVersion when using logoSvg
```
Yah I don't know on this one it isn't reproducible.  
I was just trying to explain how a tool like that could be beneficial to the site.  I tried to mention that at the top of the shared document,     It only did that one time and I made mention it could have been anything or even caching behind the scenes.  

**URL fetch an svg to json**.  I probably will add this as a function for my Github Action at some point. reduce any unnecessary leg work for developers.  I mean I doubt anyone will ever use my things but at least it will be there =) if they want it.  

**Base64 vs svg standard**

The how it happened. . .
I settled on base64 only after reading stack overflow of other developers having similar issues but it was never clear if they were using shield.io/endpoint or shields.io/badges.   Returning to the main page and seeing logo was base64 and some of the bugs listed in this forum mentioning syncing endpoint names with url named variables.  I made a bad assumption largely because base64 just made well **sense**.  It seemed much more user friendly with escape characters compared to svg standard when applied to a string embeded in json.  TBH Nothing I had read so far indicated anything else was used.  I then only filed the bug after trying several variations largely because I assumed it was something I was doing wrong,  which it was in the end.   


Again sorry for the miscommunication,  I will seriously take this experience to heart when interacting with development teams in the future.

**Comment by calebcartwright** [comment]
No worries! 

> > 1. Do not add schemaVersion when using logoSvg
>
> Yah I don't know on this one it isn't reproducible.

Excellent, it would've been concerning had that been the case.

Sounds like the only potential short-term action item on our end then would be to consider explicitly adding an `<svg>...` example to the endpoint documentation text, which we'll certainly take under advisement.

Thanks again for sharing your experience and perspective!
